File: regions/system/library/typescript/tracing/src/baggage.ts
export class Baggage {
  private entries = new Map<string, string>();

  set(key: string, value: string): void {
    this.entries.set(key, value);
  }

  get(key: string): string | undefined {
    return this.entries.get(key);
  }

  toHeader(): string {
    const parts: string[] = [];
    for (const [k, v] of this.entries) {
      parts.push(`${k}=${v}`);
    }
    return parts.join(',');
  }

  static fromHeader(s: string): Baggage {
    const baggage = new Baggage();
    if (!s) return baggage;
    for (const part of s.split(',')) {
      const eqIndex = part.indexOf('=');
      if (eqIndex > 0) {
        const key = part.substring(0, eqIndex).trim();
        const value = part.substring(eqIndex + 1).trim();
        baggage.entries.set(key, value);
      }
    }
    return baggage;
  }
}

File: regions/system/library/dart/tracing/lib/src/baggage.dart
class Baggage {
  final _entries = <String, String>{};

  void set(String key, String value) => _entries[key] = value;

  String? get(String key) => _entries[key];

  Map<String, String> get entries => Map.unmodifiable(_entries);

  String toHeader() =>
      _entries.entries.map((e) => '${e.key}=${e.value}').join(',');

  static Baggage fromHeader(String s) {
    final baggage = Baggage();
    if (s.isEmpty) return baggage;
    for (final pair in s.split(',')) {
      final idx = pair.indexOf('=');
      if (idx > 0) {
        baggage.set(pair.substring(0, idx).trim(), pair.substring(idx + 1).trim());
      }
    }
    return baggage;
  }
}
